flink-runtime/src/test/java/org/apache/flink/runtime/util/ZooKeeperUtilsTest.java

```diff
@@ -44,4 +58,27 @@ private void runZookeeperPathGenerationTest(
 
         assertThat(result, is(expectedValue));
     }
+
+    @Test
+    public void testStartCuratorFrameworkFailed() throws Exception {
+        try {
+            final SystemExitTrackingSecurityManager trackingSecurityManager =
+                    new SystemExitTrackingSecurityManager();
+            System.setSecurityManager(trackingSecurityManager);
+            Configuration configuration = new Configuration();
+            configuration.setString(HighAvailabilityOptions.HA_ZOOKEEPER_QUORUM, "localhost:2181");
+            CuratorZookeeperClient mockedCuratorZookeeperClient =
+                    PowerMockito.mock(CuratorZookeeperClient.class);
+            PowerMockito.doThrow(new RuntimeException()).when(mockedCuratorZookeeperClient).start();
+            PowerMockito.whenNew(CuratorZookeeperClient.class)
+                    .withAnyArguments()
+                    .thenReturn(mockedCuratorZookeeperClient);
```

Comment:
Instead of using PowerMockito, can we set a special `ZookeeperFactory` in the `CuratorFrameworkFactory.Builder`? I guess we would need to make the `ZooKeeperUtils` a bit easier to test (e.g. introducing a package private `startCuratorFramework(CuratorFrameworkFactory.Builder)`.

Reply:
Good point 👍